src/main/scala/cognite/spark/v1/AssetHierarchyBuilder.scala

```diff
@@ -418,9 +418,10 @@ class AssetHierarchyBuilder(config: RelationConfig)(val sqlContext: SQLContext)
     }
 }
 
-object AssetHierarchyBuilder extends NamedRelation {
+object AssetHierarchyBuilder extends NamedRelation with UpsertSchema with DeleteWithIdSchema {
```

Comment:
currently there's no compile-time link between this exposed schema and Row->schematype conversion in datasource operations so they can potentially diverge. Additionally they do seem to diverge already :sadpanda: so DeleteItemByCogniteId used by EventsRelation.delete will accept either `id` or `externalId`

A step towards linking it together could be adding more type parameters to WritableRelation and others, but there might be all the edge cases that may break or quirks that need to be kept

But overall I think it makes sense that datasource lib should hold the knowledge of schemas, this incremental step should be fine to do

Reply:
Another thought is it seems schema isn't really checked or fully used anywhere :thinking: it is exposed, there's some validation that schema exists, but compliance to schema is up to each relation and Row->schematype parsing/conversion

Reply:
Hmm, yes, these schemas could potentially be deleted eventually but they are used as-is in SchemaService. We could consider eventually reworking the jetfire side of things and maybe remove this flow completely but let's keep the scope limited